### PR TITLE
Throw WARNING when encountering non-standard file extensions for common document encoding types

### LINKED ISSUE #617: As a user, I would like to enforce browse file extension with encoding type
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Data users

### 💪 Motivation

so that the <encoding_standard_id> matches the file type in the <File_Area_Browse> 

### 📖 Additional Details

A browse label containing the following passed validation with no errors or warnings:

```
 <File_Area_Browse>
        <File>
            <file_name>uvis_euv_2005_159_solar_time_series_ingress.jpg</file_name>
            <creation_date_time>2023-03-08T14:22:14</creation_date_time>
        </File>
        <Encoded_Image>
            <offset unit="byte">0</offset>
            <encoding_standard_id>PNG</encoding_standard_id>
        </Encoded_Image>
    </File_Area_Browse>
```

### Acceptance Criteria

**Given** 
**When I perform** 
**Then I expect** 

### ⚙️ Engineering Details

Per IM Spec, it looks like we should check the file extensions for the following file types.

Possible values for `//Encoded_Image/encoding_standard_id`:
- GIF - The Encoded_Image is governed by the standard Graphics Interchange Format (GIF).
- J2C - The Encoded_Image is governed by the standard JPEG2000 compressed image codestream.
- JPEG - The Encoded_Image is governed by the standard Joint Photographic Experts Group (JPEG), ISO/IEC 10918-1.
- PDF - The Encoded_Image is governed by the standard Portable Document Format (PDF), ISO 32000-1:2008.
- PDF/A - The Encoded_Image is governed by the standard Portable Document Format / Archive (PDF/A), ISO 19005-1:2005.
- PNG - The Encoded_Image is governed by the standard Portable Network Graphics (PNG), ISO/IEC 15948:2004.
- TIFF - The Encoded_Image is governed by the standard Tagged Image File Format (TIFF), Version 6.n, 1992.

### PR DESCRIPTION
**Pull request by al-niessner**
## 🗒️ Summary
Added a check that compares the file extension (suffix of filename from last .) to the those known or allowed for the given encoding type, When it fails it generates the following message:

```
WARNING  [error.validation.file_naming_problem]   From the encoding type 'PNG'the file extension 'jpg' is not one of the allowed: [png]
```

## ⚙️ Test Data and/or Report
See unit testing below

## ♻️ Related Issues
Closes #617

### CODE REVIEW

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java [hunk 1]
```diff
@@ -200,32 +210,40 @@ private boolean validate(DocumentInfo xml) {
...
             String documentStandardId = null;
-            List<TinyNodeImpl> children = new ArrayList<>();
+            String encodingStandardId = null;
+            TinyNodeImpl fileObject = null;
+            List<TinyNodeImpl> grandChildren = new ArrayList<>();
             try {
-              children = extractor.getNodesFromItem("*", fileObject);
+              List<TinyNodeImpl> children = extractor.getNodesFromItem("*", fileAreaObject);
+              for (TinyNodeImpl child : children) {
+                if ("File".equals(child.getLocalPart()) || "Document_File".equals(child)) fileObject = child;
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java]: @sonatype-lift ignore

This way is null safe

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java [hunk 2]
```diff
@@ -272,8 +290,13 @@ private boolean validate(DocumentInfo xml) {
                 filesize = child.getStringValue();
                 LOG.debug("FileReferenceValidationRule:validate:filesize {}", filesize);
               }
+              if ("encoding_standard_id".equals(child.getLocalPart())) {
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java]: @sonatype-lift ignore

This way is null safe

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java [hunk 3]
```diff
@@ -200,31 +208,41 @@ private boolean validate(DocumentInfo xml) {
...
             String directory = "";
             String filesize = "";
             String documentStandardId = null;
-            List<TinyNodeImpl> children = new ArrayList<>();
+            String encodingStandardId = null;
+            TinyNodeImpl fileObject = null;
+            List<TinyNodeImpl> grandChildren = new ArrayList<>();
             try {
-              children = extractor.getNodesFromItem("*", fileObject);
+              if ("Document_File".equals(fileAreaObject.getLocalPart())) {
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java:243]: @sonatype-lift ignore

This way is null safe

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java [hunk 4]
```diff
@@ -200,31 +208,41 @@ private boolean validate(DocumentInfo xml) {
...
+                grandChildren = extractor.getNodesFromItem("*", fileAreaObject);
+              } else {
+                List<TinyNodeImpl> children = extractor.getNodesFromItem("*", fileAreaObject);
+                for (TinyNodeImpl child : children) {
+                  if ("File".equals(child.getLocalPart())) fileObject = child;
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java:249]: @sonatype-lift ignore

This way is null safe

File: src/main/java/gov/nasa/pds/tools/util/EncodingMimeMapping.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+public enum EncodingMimeMapping {
+  GIF(Arrays.asList("gif")),
+  J2C(Arrays.asList("j2c", "mj2", "mjp2")),
+  JPEG(Arrays.asList("jpg", "jpeg")),
+  PDF(Arrays.asList("pdf")),
+  PDFA(Arrays.asList("pdf", "pdfa")),
+  PNG(Arrays.asList("png")),
+  TIFF(Arrays.asList("tif", "tiff"));
+  final private List<String> extensions;
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/util/EncodingMimeMapping.java:15]: @sonatype-lift ignore

Given class design and implementation, this is a false positive. The list is final and private so it cannot be changed wholesale by the class. Only copies of it are shared externally. Hence it is immutable.

**Comment by al-niessner** [comment]
@jordanpadams @nutjob4life @tloubrieu-jpl 

Ready for review. Can sonatype be run again?

**Review by nutjob4life — APPROVED** [review]
👍

**Pull request merged**